src/udev/scsi_id/scsi_serial.c

```diff
@@ -85,8 +89,111 @@ static const char hex_str[]="0123456789abcdef";
 #define SG_ERR_CAT_SENSE               98        /* Something else in the sense buffer */
 #define SG_ERR_CAT_OTHER               99        /* Some other error/warning */
 
-static int do_scsi_page80_inquiry(struct scsi_id_device *dev_scsi, int fd,
-                                  char *serial, char *serial_short, int max_len);
+#define PAGE_PREFIX "/sys/block/"
+#define PAGE0x00_SUFFIX "/device/inquiry"
+#define PAGE0x80_SUFFIX "/device/vpd_pg80"
+#define PAGE0x83_SUFFIX "/device/vpd_pg83"
+
+static int process_page0_inquiry_data(struct scsi_id_device *dev_scsi,
+                                      unsigned char *buffer, unsigned int len);
+
+static int process_page83_inquiry_data(struct scsi_id_device *dev_scsi, char *serial,
+                                       char *serial_short, unsigned char *page_83, int len, char *unit_serial_number,
+                                       char *wwn, char *wwn_vendor_extension, char *tgpt_group);
+
+static int process_page80_inquiry_data(struct scsi_id_device *dev_scsi, char *serial,
+                                       char *serial_short, unsigned char *buf, int max_len);
+
+static int process_page83_prespc3_inquiry_data(struct scsi_id_device *dev_scsi,
+                                               char *serial, char *serial_short, unsigned char *page_83,
+                                               int len);
+static void process_std_inquiry_data(struct scsi_id_device *dev_scsi, unsigned char *buf);
+
+static int sgio_do_scsi_page80_inquiry(struct scsi_id_device *dev_scsi, int fd,
+                                       char *serial, char *serial_short, int max_len);
+static void scsi_get_dev_name(const char *filePath, char *dev_name);
+static void scsi_get_device_vpd_page_path(const char *device_path, char *device_vpd_page_path, unsigned int page);
+
+static void scsi_get_dev_name(const char *file_path, char *dev_name) {
+        _cleanup_free_ char *file_name = NULL;
+        int r;
+
+        r = path_extract_filename(file_path, &file_name);
+
+        if (r < 0 || r == O_DIRECTORY) {
+                printf("Invalid Path Specified\n");
+                exit(0);
+        }
+        strcpy(dev_name, file_name);
+}
+
+static void scsi_get_device_vpd_page_path(const char *device_path, char *device_vpd_page_path, unsigned int page) {
+        char dev_name[MAX_ATTR_LEN];
```

Comment:
Then make this `_cleanup_free_ char *dev_name = NULL;`

Reply:
Done